### PR TITLE
[v2] Use the v2 components runtime as the core of the Elastic Agent

### PR DESCRIPTION
**Pull request by blakerouse**
## What does this PR do?

Changes the core management of the Elastic Agent to use the `coordinator` that manages the interaction between the configuration manager, runtime manager, and composable variable manager.

At the moment the unit tests will fail, there is still more to do, and more code to remove (like the operators, programs, old specifications, etc.). Some code was already removed as it help to find areas that needed to be fixed and just made the job easier to remove them.

This is being proposed because its already a large chunk of changes and it would be better to get it in there now vs keeping it local to me.

## Why is it important?

Work towards switching to new v2 design of the Elastic Agent.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- ~~[ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.~~

### CODE REVIEW

**Review by michel-laterman — COMMENTED** [review]
tried to take a look, I think I would benefit from a short walkthrough on what is being moved/replaced (i.e., where do fleet interactions occur now)

File: internal/pkg/agent/application/actions/handlers/handler_action_policy_change.go [hunk 1]
```diff
@@ -210,3 +218,33 @@ func fleetToReader(agentInfo *info.AgentInfo, cfg *configuration.Configuration)
...
+	acker  acker.Acker
+	commit bool
+}
+
+func (l *policyChange) Config() *config.Config {
+	return l.cfg
+}
+
+func (l *policyChange) Ack() error {
+	if l.action == nil {
```
[the remark this answers is not in the document]
Comment (blakerouse) [reply internal/pkg/agent/application/actions/handlers/handler_action_policy_change.go:235]: Yeah I thought the lint there was very weird. Going to ignore it in the PR, but we should look into fixing it.

File: internal/pkg/agent/application/upgrade/upgrade.go [hunk 2]
```diff
@@ -112,40 +82,17 @@ func (u *Upgrader) Upgradeable() bool {
 
 // Upgrade upgrades running agent, function returns shutdown callback if some needs to be executed for cases when
 // reexec is called by caller.
-func (u *Upgrader) Upgrade(ctx context.Context, a Action, reexecNow bool) (_ reexec.ShutdownCallbackFn, err error) {
+func (u *Upgrader) Upgrade(ctx context.Context, version string, sourceURI string, action *fleetapi.ActionUpgrade) (_ reexec.ShutdownCallbackFn, err error) {
```
[the remark this answers is not in the document]
Comment (blakerouse) [reply internal/pkg/agent/application/upgrade/upgrade.go:85]: Yes it gets passed in from the Upgrade action handler from the Fleet Gateway, as well as from the control protocol.

File: internal/pkg/agent/application/upgrade/upgrade.go [hunk 3]
```diff
@@ -155,13 +102,7 @@ func (u *Upgrader) Upgrade(ctx context.Context, a Action, reexecNow bool) (_ ree
 	}
 
 	if strings.HasPrefix(release.Commit(), newHash) {
-		// not an error
-		if action := a.FleetAction(); action != nil {
-			//nolint:errcheck // keeping the same behavior, and making linter happy
-			u.ackAction(ctx, action)
-		}
-		u.log.Warn("upgrading to same version")
```
[the remark this answers is not in the document]
Comment (blakerouse) [reply internal/pkg/agent/application/upgrade/upgrade.go:163]: It should not happen ever.

File: internal/pkg/agent/cmd/status.go [hunk 4]
```diff
@@ -73,45 +73,45 @@ func statusCmd(streams *cli.IOStreams, cmd *cobra.Command, args []string) error
...
+	if len(state.Components) == 0 {
+		fmt.Fprint(w, "Components: (none)\n")
 	} else {
-		fmt.Fprint(w, "Applications:\n")
+		fmt.Fprint(w, "Components:\n")
 		tw := tabwriter.NewWriter(w, 4, 1, 2, ' ', 0)
-		for _, app := range status.Applications {
-			fmt.Fprintf(tw, "  * %s\t(%s)\n", app.Name, app.Status)
-			if app.Message == "" {
+		for _, comp := range state.Components {
```
[the remark this answers is not in the document]
Comment (blakerouse) [reply internal/pkg/agent/cmd/status.go:109]: I do want to greatly improve this output. It is something that needs to be done in a follow up. At the moment I got it working just enough for the status output to provide data. JSON and YAML output provide all the detail at the moment.

I want to add the ability to do `elastic-agent status --watch` that will use something like tcell to provide a stream of status changes for the agent, components, and units.

File: internal/pkg/agent/configuration/grpc.go [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+// GRPCConfig is a configuration of GRPC server.
+type GRPCConfig struct {
+	Address string `config:"address"`
+	Port    uint16 `config:"port"`
+}
+
+// DefaultGRPCConfig creates a default server configuration.
+func DefaultGRPCConfig() *GRPCConfig {
+	return &GRPCConfig{
+		Address: "localhost",
```
[the remark this answers is not in the document]
Comment (blakerouse) [reply internal/pkg/agent/configuration/grpc.go:18]: Yes we do. Would like to do that before the release of v2, but only once we get v2 working.

File: internal/pkg/composable/controller.go [hunk 6]
```diff
@@ -87,28 +93,40 @@ func New(log *logger.Logger, c *config.Config) (Controller, error) {
...
-			return errors.New(err, fmt.Sprintf("failed to run provider '%s'", name), errors.TypeConfig, errors.M("provider", name))
-		}
+		go func() {
+			defer wg.Done()
+			err := state.provider.Run(state)
+			if err != nil && !errors.Is(err, context.Canceled) {
+				err = errors.New(err, fmt.Sprintf("failed to run provider '%s'", name), errors.TypeConfig, errors.M("provider", name))
+				c.logger.Errorf("%s", err)
+			}
+		}()
```
[the remark this answers is not in the document]
Comment (blakerouse) [reply internal/pkg/composable/controller.go:129]: I think you are correct here, seems lint below caught it but not this one. I fixed both.

File: internal/pkg/fleetapi/acker/noop/noop_acker.go [hunk 7]
```diff
@@ -2,27 +2,29 @@
 // or more contributor license agreements. Licensed under the Elastic License;
 // you may not use this file except in compliance with the Elastic License.
 
-package fleet
+package noop
```
[the remark this answers is not in the document]
Comment (blakerouse) [reply internal/pkg/fleetapi/acker/noop/noop_acker.go:5]: It used to be used not just in testing, but with the changes in this branch it is now only used by tests.

I think we could clean that up in a followup. Its a very small file and it hurting anything where it is sitting at the moment.

File: pkg/component/runtime/manager.go [hunk 8]
```diff
@@ -48,6 +48,12 @@ var (
 	ErrNoUnit = errors.New("no unit under control of this manager")
 )
 
+// ComponentComponentState provides a structure to map a component to current component state.
+type ComponentComponentState struct {
```
[the remark this answers is not in the document]
Comment (blakerouse) [reply pkg/component/runtime/manager.go:52]: Open to suggestions, I was not happy with it either but couldn't think of a better name.

File: pkg/component/runtime/manager.go [hunk 9]
```diff
@@ -290,26 +323,100 @@ func (m *Manager) PerformAction(ctx context.Context, unit component.Unit, name s
...
 	m.mx.RUnlock()
 	if ok {
 		comp.latestMx.RLock()
-		sub.ch <- comp.latestState
+		latestState := comp.latestState.Copy()
 		comp.latestMx.RUnlock()
+		go func() {
+			select {
+			case <-ctx.Done():
+			case sub.ch <- latestState:
```
[the remark this answers is not in the document]
Comment (blakerouse) [reply pkg/component/runtime/manager.go:343]: Yes.

File: version/version.go [hunk 10]
```diff
@@ -4,4 +4,4 @@
 
 package version
 
-const defaultBeatVersion = "8.4.0"
+const defaultBeatVersion = "8.3.0"
```
[the remark this answers is not in the document]
Comment (blakerouse) [reply version/version.go:7]: No it was not. I changed it back. I only did that to test my locally built agent with Elastic Cloud 8.3.

File: pkg/component/runtime/runtime.go [hunk 11]
```diff
@@ -35,12 +35,24 @@ type ComponentUnitKey struct {
 	UnitID   string
 }
 
+// ComponentVersionInfo provides version information reported by the component.
+type ComponentVersionInfo struct {
```
[the remark this answers is not in the document]
Comment (blakerouse) [reply pkg/component/runtime/runtime.go:39]: Benefit of v2 protocol!

File: internal/pkg/agent/application/coordinator/coordinator.go [hunk 12]
```diff
@@ -0,0 +1,592 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package coordinator
```
[the remark this answers is not in the document]
Comment (blakerouse) [reply internal/pkg/agent/application/coordinator/coordinator.go:5]: Whats the benefit for package doc string? I added doc strings to all public structs, interfaces, and functions. Not clear to me what you would want here.

It's also not something done across the code base at the moment. Is that something we want to start doing for the entire code base?

**Review by blakerouse — COMMENTED** [review]
Happy to get on a call and walk through any part of the PR.

**Review by michel-laterman — APPROVED** [review]
👍 looking forward to the deletions!

File: internal/pkg/agent/cmd/diagnostics.go [hunk 13]
```diff
@@ -428,29 +428,31 @@ func gatherConfig() (AgentConfig, error) {
...
+			return AgentConfig{}, err
 		}
-	}
+		cfg.AppConfig = make(map[string]interface{}, 0)
+		for rk, programs := range pMap {
+			for _, p := range programs {
+				cfg.AppConfig[p.Identifier()+"_"+rk] = p.Configuration()
+			}
+		}
+	*/
```
Comment (ph) [inline internal/pkg/agent/cmd/diagnostics.go:456]: disabled to reenable it later right @blakerouse?

File: internal/pkg/agent/configuration/grpc.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ph) [reply internal/pkg/agent/configuration/grpc.go:18]: Yes the idea is to remove the network overhead and the possibility of port conflicts.

File: pkg/component/runtime/manager.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (ph) [reply pkg/component/runtime/manager.go:52]: `ExecutionState` `ComponentExecutionState` ?

**Comment by ph** [comment]
@blakerouse Good job on this, you merge it before I finished the review, nothing major on my side added a few suggestions.

**Pull request merged**